File: client/src/shared/api/analytics.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080/api";

export interface AnalyticsData {
  total_visits: number;
  unique_visitors: number;
  today_visits: number;
  week_visits: number;
  avg_session_duration: number;
  bounce_rate: number;
  page_views: Record<string, number>;
  top_pages: { page: string; views: number }[];
  visits_by_day: { date: string; visits: number }[];
  visits_by_hour: number[];
  devices: { desktop: number; mobile: number; tablet: number };
  themes: { light: number; dark: number };
  languages: { ru: number; en: number };
}

// Track page view
export async function trackPageView(page: string, visitorId: string, device: string) {
  try {
    await fetch(`${API_URL}/analytics/track`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        event: "page_view",
        page,
        visitor_id: visitorId,
        device,
      }),
    });
  } catch {
    // Silently fail - analytics shouldn't break the app
  }
}

// Track session end
export async function trackSession(
  visitorId: string,
  duration: number,
  pages: number,
  theme: string,
  language: string
) {
  try {
    await fetch(`${API_URL}/analytics/track`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        event: "session",
        visitor_id: visitorId,
        duration,
        pages,
        theme,
        language,
      }),
    });
  } catch {
    // Silently fail
  }
}

// Get analytics (admin only)
export async function getAnalytics(password: string): Promise<AnalyticsData | null> {
  try {
    const res = await fetch(`${API_URL}/analytics`, {
      headers: { "X-Admin-Password": password },
    });
    if (!res.ok) return null;
    return res.json();
  } catch {
    return null;
  }
}

File: server/internal/entity/analytics.go
package entity

type PageView struct {
	ID        int64  `json:"id"`
	Page      string `json:"page"`
	VisitorID string `json:"visitor_id"`
	Device    string `json:"device"`
	CreatedAt string `json:"created_at"`
}

type Session struct {
	ID              int64  `json:"id"`
	VisitorID       string `json:"visitor_id"`
	DurationSeconds int    `json:"duration_seconds"`
	PagesCount      int    `json:"pages_count"`
	Theme           string `json:"theme"`
	Language        string `json:"language"`
	CreatedAt       string `json:"created_at"`
	UpdatedAt       string `json:"updated_at"`
}

type DailyStats struct {
	ID             int64  `json:"id"`
	Date           string `json:"date"`
	Visits         int    `json:"visits"`
	UniqueVisitors int    `json:"unique_visitors"`
	DesktopCount   int    `json:"desktop_count"`
	MobileCount    int    `json:"mobile_count"`
	TabletCount    int    `json:"tablet_count"`
	LightTheme     int    `json:"light_theme"`
	DarkTheme      int    `json:"dark_theme"`
	LangRu         int    `json:"lang_ru"`
	LangEn         int    `json:"lang_en"`
}

type AnalyticsData struct {
	TotalVisits        int                    `json:"total_visits"`
	UniqueVisitors     int                    `json:"unique_visitors"`
	TodayVisits        int                    `json:"today_visits"`
	WeekVisits         int                    `json:"week_visits"`
	AvgSessionDuration float64                `json:"avg_session_duration"`
	BounceRate         float64                `json:"bounce_rate"`
	PageViews          map[string]int         `json:"page_views"`
	TopPages           []TopPage              `json:"top_pages"`
	VisitsByDay        []DayVisits            `json:"visits_by_day"`
	VisitsByHour       []int                  `json:"visits_by_hour"`
	Devices            DeviceStats            `json:"devices"`
	Themes             ThemeStats             `json:"themes"`
	Languages          LanguageStats          `json:"languages"`
}

type TopPage struct {
	Page  string `json:"page"`
	Views int    `json:"views"`
}

type DayVisits struct {
	Date   string `json:"date"`
	Visits int    `json:"visits"`
}

type DeviceStats struct {
	Desktop int `json:"desktop"`
	Mobile  int `json:"mobile"`
	Tablet  int `json:"tablet"`
}

type ThemeStats struct {
	Light int `json:"light"`
	Dark  int `json:"dark"`
}

type LanguageStats struct {
	Ru int `json:"ru"`
	En int `json:"en"`
}

type TrackEventRequest struct {
	Event     string `json:"event"`
	Page      string `json:"page,omitempty"`
	VisitorID string `json:"visitor_id"`
	Device    string `json:"device,omitempty"`
	Theme     string `json:"theme,omitempty"`
	Language  string `json:"language,omitempty"`
	Duration  int    `json:"duration,omitempty"`
	Pages     int    `json:"pages,omitempty"`
}
